NXP-19645: Use different CUDF solver criteria for each behaviour

yjulienne-nuxeo opened the pull request.

yjulienne-nuxeo committed:
NXBT-652: Use different CUDF solver criteria for each behaviour

jcarsique commented:
You must use an NXBT JIRA ticket for changes related to nuxeo-connect (https://jira.nuxeo.com/browse/NXBT/component/12900)

jcarsique commented on nuxeo-connect-client/src/main/java/org/nuxeo/connect/packages/dependencies/DependencyResolver.java:
```diff
@@ -57,4 +57,29 @@ public DependencyResolution resolve(List<String> pkgInstall,
+     * are not sure of what to give as a solverCriteria
+     *
+     * @param solverCriteria specify the criteria string to be used by the solver
+     * @since TODO
```
replace TODO

jcarsique commented on nuxeo-connect-client/src/main/java/org/nuxeo/connect/packages/dependencies/DependencyResolver.java:
```diff
@@ -57,4 +57,29 @@ public DependencyResolution resolve(List<String> pkgInstall,
+     * @param allowSNAPSHOT true to allow SNAPSHOT packages to be part of the computed solution
+     * @param doKeep false to uninstall all packages that do not need to be part of the computed solution
+     * @param solverCriteria specify the criteria string to be used by the solver
+     * @since TODO
```
TODO

jcarsique commented on nuxeo-connect-client/src/main/java/org/nuxeo/connect/packages/dependencies/LegacyDependencyResolver.java:
```diff
@@ -149,4 +149,17 @@ public DependencyResolution resolve(List<String> pkgInstall,
+    @Override
+    public DependencyResolution resolve(List<String> pkgInstall, List<String> pkgRemove, List<String> pkgUpgrade,
+            String targetPlatform, String solverCriteria) throws DependencyException {
+        throw new UnsupportedOperationException("Legacy resolver does not support advanced resolution method");
```
This is fine and can stay as is. An alternative would be to remove the LegacyDependencyResolver which is deprecated since 1.4.13 (no more used by any maintained Nuxeo platform).

jcarsique commented on nuxeo-connect-client/src/main/java/org/nuxeo/connect/packages/dependencies/P2CUDFDependencyResolver.java:
```diff
@@ -44,6 +45,22 @@
+     * - mp-remove  -> SOLVER_CRITERIA_LESS_VERSION_CHANGES
+     * - mp-upgrade -> SOLVER_CRITERIA_LESS_OUTDATED
+     * - mp-set     -> SOLVER_CRITERIA_LESS_OUTDATED_WITH_REMOVE
+     * @since TODO
```
Invalid Javadoc format: use HTML list and place it in P2CUDFDependencyResolver doc. Placed here, it will apply to `SOLVER_CRITERIA_LESS_CHANGES`. Add also a short single Javadoc line for each constant like "Solution with less package changes"
Replace TODO

jcarsique commented on nuxeo-connect-client/src/main/java/org/nuxeo/connect/packages/dependencies/P2CUDFDependencyResolver.java:
```diff
@@ -75,9 +92,40 @@ public DependencyResolution resolve(List<String> pkgInstall,
+            solverCriteria = SOLVER_CRITERIA_LESS_OUTDATED;
+        } else if (CollectionUtils.isNotEmpty(pkgRemove)) {
+            // For a remove request, criteria prioritizes the solution with the less version changed packages
+            // otherwise, it will upgrade/downgrade a package instead of removing it when trying to remove a specific
```
it will / it would

jcarsique commented on nuxeo-connect-client/src/main/java/org/nuxeo/connect/packages/dependencies/P2CUDFDependencyResolver.java:
```diff
@@ -159,19 +206,19 @@ public DependencyResolution resolve(List<String> pkgInstall,
+     * return a no-change solution
+     *
+     * @deprecated since 1.4.26 Use {@link #resolve(List, List, List, String)} instead
+     */
```
<< This method was used for backward compatibility with the {@link LegacyDependencyResolver} API. It is not possible with the current implementation (1.4.26) to upgrade a single package without specifying its version, so calling this method on an already installed package will return a no-change solution. >>

Set (move) the deprecation on the `DependencyResolver` interface.
Remove its usage cases.

yjulienne-nuxeo replied:
What do you mean by "Remove its usage cases" ? Replace by the good method in all tests ?

jcarsique commented on nuxeo-connect-client/src/test/java/org/nuxeo/connect/pm/tests/TestPackageDependencyResolutionP2CUDF.java:
```diff
@@ -57,59 +57,67 @@ protected void tearDown() throws Exception {
         CUDFHelper.defaultAllowSNAPSHOT = false;
     }
 
+    // Before: [AA-1.0.0, BB-1.0.0, Blocker-1.0.0, Blocker2-1.0.0, EE-1.0.0, NXBT-654.1-1.0.0, NXBT-654.2-1.0.0,
```
This class makes use of deprecated `DummyPackageSource` constructor. Since it's an easy fix, it is recommended (best practice) to take the chance and reduce the technical debt while editing that file.

jcarsique reviewed (changes requested):
Format your code (within a separate commit).

yjulienne-nuxeo committed:
NXBT-652: Cleanup and format

yjulienne-nuxeo committed:
NXBT-652: Remove usages of deprecated resolve(String, String) method.

And deprecate calling PackageManager.resolveDependencies(String, String)
method.
